src/tools/ceph_dedup_tool.cc

```diff
@@ -562,34 +569,68 @@ class SampleDedupWorkerThread : public Thread
       std::unique_lock lock(fingerprint_lock);
       auto found_iter = fp_map.find(chunk.fingerprint);
       ssize_t cur_reference = 1;
+      if (is_hit_threshold()) {
+	auto len_before = fp_map.size();
+	try_shrink();
+	auto len_after = fp_map.size();
+	if (len_before - len_after < len_before / 4) {
+	  state = fp_store_state_t::FULL;
+	  cout << "FpStore reaches memory threshold" << std::endl;
+	}
+      }
       if (found_iter == fp_map.end()) {
-        fp_map.insert({chunk.fingerprint, 1});
+	if (state != fp_store_state_t::FULL) {
+	  fp_map.insert({chunk.fingerprint, 1});
+	}
       } else {
 	cur_reference = ++found_iter->second;
       }
       return cur_reference >= dedup_threshold && dedup_threshold != -1;
     }
 
+    bool is_hit_threshold() {
```

Comment:
This can be simpler.

```
size_t get_fp_map_size() const {
  return fp_map.size() * (fp_map.begin()->first.length() + sizeof(dup_count_t));
}
bool is_fpmap_full() const {
  return get_fp_map_size() >= memory_threshhold;
}
```

Reply:
Done.